File: wrappers/uniffi/swift/Tests/AsdkTests/Kms.swift
import Testing
import Foundation
import Swifter
@testable import Asdk

@Suite(.serialized) class KmsTests {


    @Test func create() async throws {
        let kms = self.mockKms()
        let kid = try await kms.create(kt: .ed25519)

        #expect("TestString" == kid) 
    }

    @Test func getByKeyId() async throws {
        let kms = self.mockKms()
        let keyHandle = try await kms.get(kid: "TestString")

        #expect(try keyHandle.inner.pubKey() == "testKey".data(using: .utf8)!)
    }

    @Test func getByPublicKey() async throws {
        let publicKey = "testKey".data(using: .utf8)!
        let kms = self.mockKms()
        let keyHandle = try await kms.getByPublicKey(publicKey: publicKey)

        #expect(try keyHandle.inner.pubKey() == "testKey".data(using: .utf8)!)
    }
    
    private func mockKms() -> Asdk.Kms {
        return wrapKmsForTests(kms: MockKms())
    }

}


final class MockKms : Asdk.Kms {
    func create(kt: Asdk.KeyType) async throws -> String {
        return "TestString"
    }
    
    func get(kid: String) async throws -> Asdk.WrappedKeyHandle {
        return Asdk.WrappedKeyHandle(inner: MockKeyHandle())
    }
    
    func getByPublicKey(publicKey: Data) async throws -> Asdk.WrappedKeyHandle {
        return Asdk.WrappedKeyHandle(inner: MockKeyHandle())
    }
    
}

File: wrappers/uniffi/kotlin/src/test/kotlin/com/bci/Kms.kt
package com.bci

import com.bci.asdk.KeyType
import com.bci.asdk.WrappedKeyHandle
import com.bci.asdk.setJniLibPath
import com.bci.asdk.wrapKmsForTests
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.BeforeAll
import kotlin.test.Test
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import com.bci.asdk.Kms as ASDKKms

class Kms {
    companion object {
        @JvmStatic
        @BeforeAll
        fun setup() {
            setJniLibPath()
        }
    }

    @Test
    fun create() = runTest {
        val kms = mockKms()
        val kid = kms.create(KeyType.P256)

        assertEquals(kid, "TestString")
    }

    @Test
    fun getByKeyId() = runTest {
        val kms = mockKms()
        val keyHandle = kms.get("TestString")

        assertEquals("testKey", String(keyHandle.inner.pubKey(), Charsets.UTF_8))
    }

    @Test
    fun getByPublicKey() = runTest {
        val kms = mockKms()
        val keyHandle = kms.getByPublicKey("testKey".toByteArray())

        assertEquals("testKey", String(keyHandle.inner.pubKey(), Charsets.UTF_8))
    }

    private fun mockKms(): ASDKKms {
        return wrapKmsForTests(MockKms())
    }
}


class MockKms : ASDKKms {
    override suspend fun create(kt: KeyType): String {
        return "TestString"
    }

    override suspend fun get(kid: String): WrappedKeyHandle {
        return WrappedKeyHandle(MockKeyHandle())
    }

    override suspend fun getByPublicKey(publicKey: ByteArray): WrappedKeyHandle {
        return WrappedKeyHandle(MockKeyHandle())
    }

}